File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        l = 0
        for c in s:
            if c == '*':
                if l > 0:
                    l -= 1
            elif c == '#':
                l *= 2
            elif c == '%':
                continue
            else:
                l += 1
        
        if k > l - 1:
            return '.'

        for i in range(len(s) - 1, -1, -1):
            if s[i] == '*':
                l += 1
            elif s[i] == '#':
                l //= 2
                if k >= l:
                    k -= l
            elif s[i] == '%':
                k = l - k - 1
            else:
                if k == l - 1:
                    return s[i]
                else:
                    l -= 1
        
        return '.'

# simulation, string
# time: O(n)
# space: O(1)

File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.go
func processStr(s string, k int64) byte {
    // find the result's length
	l := int64(0)
    for i := 0; i < len(s); i++ {
        if s[i] == '*' {
            if l > 0 {
                l--
            }
        } else if s[i] == '#' {
            l *= 2
        } else if s[i] == '%' {
            continue
        } else {
            l++
        }
    }

    if k > l - 1 {
        return '.'
    }

    // process the s backward, reverse-engineer the k
    for i := len(s) - 1; i >= 0; i-- {
        if s[i] == '*' {
            l++
        } else if s[i] == '#' {
            l /= 2
            if k >= l { 
                k = k - l
            }
        } else if s[i] == '%' {
            k = l - k - 1
        } else {
            if k == l - 1 {
                return s[i]
            } else {
                l--
            }
        }
    }

    return '.'
}

// simulation, string
// time: O(n)
// space: O(1)